src/swarm-tui/Swarm/TUI/Model/KeyBindings.hs

```diff
@@ -0,0 +1,105 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- |
+-- SPDX-License-Identifier: BSD-3-Clause
+module Swarm.TUI.Model.KeyBindings (
+  initKeyHandlingState,
+  KeybindingPrint (..),
+  showKeybindings,
+) where
+
+import Brick.Keybindings as BK
+import Control.Carrier.Lift (runM)
+import Control.Carrier.Throw.Either (runThrow)
+import Control.Effect.Accum
+import Control.Effect.Lift
+import Control.Effect.Throw
+import Control.Lens hiding (from, (<.>))
+import Data.Maybe (fromMaybe, mapMaybe)
+import Data.Text (Text)
+import Data.Text qualified as T
+import Swarm.Game.Failure (Asset (..), LoadingFailure (..), SystemFailure (..))
+import Swarm.Game.ResourceLoading (getSwarmConfigIniFile)
+import Swarm.Language.Pretty (prettyText)
+import Swarm.TUI.Controller.EventHandlers
+import Swarm.TUI.Model
+import Swarm.TUI.Model.Event (SwarmEvent, defaultSwarmBindings, swarmEvents)
+
+loadKeybindingConfig ::
+  (Has (Throw SystemFailure) sig m, Has (Lift IO) sig m) =>
+  m [(SwarmEvent, BindingState)]
+loadKeybindingConfig = do
+  (iniExists, ini) <- sendIO getSwarmConfigIniFile
+  if not iniExists
+    then return []
+    else do
+      loadedCustomBindings <- sendIO $ keybindingsFromFile swarmEvents "keybindings" ini
+      case loadedCustomBindings of
+        Left e -> throwError $ AssetNotLoaded Keybindings ini (CustomMessage $ T.pack e)
+        Right bs -> pure $ fromMaybe [] bs
+
+initKeyHandlingState ::
+  (Has (Throw SystemFailure) sig m, Has (Lift IO) sig m) =>
+  m KeyEventHandlingState
+initKeyHandlingState = do
+  customBindings <- loadKeybindingConfig
+  let cfg = newKeyConfig swarmEvents defaultSwarmBindings customBindings
+  dispatchers <- createKeyDispatchers cfg
+  return $ KeyEventHandlingState cfg dispatchers
+
+data KeybindingPrint = MarkdownPrint | TextPrint | IniPrint
+  deriving (Eq, Ord, Show)
+
+showKeybindings :: KeybindingPrint -> IO Text
+showKeybindings kPrint = do
+  bindings <- runM $ runThrow @SystemFailure initKeyHandlingState
+  pure $ case bindings of
+    Left e -> prettyText e
+    Right bs -> showTable kPrint (bs ^. keyConfig) sections
+ where
+  showTable = \case
+    MarkdownPrint -> keybindingMarkdownTable
+    TextPrint -> keybindingTextTable
+    IniPrint -> keybindingINI
+  sections =
+    [ ("main", mainEventHandlers)
+    , ("repl", replEventHandlers)
+    , ("world", worldEventHandlers)
+    , ("robot", robotEventHandlers)
+    ]
+
+keybindingINI :: Ord k => KeyConfig k -> [(Text, [KeyEventHandler k m])] -> Text
+keybindingINI kc sections =
+  T.unlines $
+    "[keybindings]\n"
+      : "; Uncomment the assignment and set comma separated list"
+      : "; of keybindings or \"unbind\" on the right. See:"
+      : "; https://hackage.haskell.org/package/brick/docs/Brick-Keybindings-Parse.html#v:parseBinding\n"
+      : map (keyBindingEventINI kc) handlersData
+ where
+  handlersData = concatMap (mapMaybe handlerData . snd) sections
```

Comment:
Could we print some section headers in comments in the generated `.ini` file as well?  Looking at the generated file it's impossible to know which keybindings are active in which substates.  For example there is a line that says `show fps = f` but there's no indication that only works when the world panel is focused.  I'm imagining a header something like `;;; World keybindings (only active when the world panel is focused) ;;;`.

Reply:
Sure, I was just lazy when copy-pasting parts from the markdown printer implementation in Brick. 🙂